compiler/modules/bir/verify.bal

```diff
@@ -79,6 +92,73 @@ type Error err:Semantic|err:Internal;
 function verifyBasicBlock(VerifyContext vc, BasicBlock bb) returns Error? {
     foreach Insn insn in bb.insns {
         check verifyInsn(vc, insn);
+        check verifyInsnRegisterKinds(vc, insn);
+    }
+}
+
+type MultipleOperandInsn IntBinaryInsn|IntNoPanicArithmeticBinaryInsn|FloatArithmeticBinaryInsn
+    |DecimalArithmeticBinaryInsn|CompareInsn|ListConstructInsn|ListGetInsn|MappingConstructInsn
+    |MappingGetInsn|StringConcatInsn|EqualityInsn|TypeMergeInsn;
+
+type ResultInsn readonly & record { *ResultInsnBase; };
+
+function verifyInsnRegisterKinds(VerifyContext vc, Insn insn) returns Error? {
+    if insn is ResultInsn {
+        if (vc.getNumTmpRegisters() > insn.result.number) && vc.getUsedTmpRegisters(insn.result.number) {
+            return vc.invalidErr("tmp register defined in multiple places", <Position>insn.result.pos);
+        }
+        else {
+            vc.addTmpRegisters(insn.result.number);
+        }
+    }
+
+    if insn is BranchInsn|CatchInsn {
+        return;
+    }
+    else if insn is CallInsn {
+        foreach Operand arg in insn.args {
+            check verifyRegisterKind(vc, arg);
+        }
+    }
+    else if insn is ListSetInsn|MappingSetInsn {
+        check verifyNonFinalRegisterKind(vc, insn.operands[0]);
+        check verifyRegisterKind(vc, insn.operands[1]);
+        check verifyRegisterKind(vc, insn.operands[2]);
+    }
+    else if insn is MultipleOperandInsn {
+        // JBUG #35557 can't iterate insn.operands
+        Operand[] operands = insn.operands;
+        foreach Operand op in operands {
+            check verifyRegisterKind(vc, op);
+        }
+    }
+    else {
+        check verifyRegisterKind(vc, insn.operand);
+    }
+}
+
+function verifyNonFinalRegisterKind(VerifyContext vc, Operand r) returns Error? {
+    if r is FinalRegister {
+        return vc.invalidErr("invalid register kind final for insn: " + r.kind, <Position>r.pos);
+    }
+    check verifyRegisterKind(vc, r);
+}
+
+function verifyRegisterKind(VerifyContext vc, Operand r) returns Error? {
+    if r !is Register {
+        return;
+    }
+    if r is TmpRegister && (vc.getNumTmpRegisters() <= r.number || !vc.getUsedTmpRegisters(r.number)) {
+        return vc.invalidErr("tmp register not initialized", <Position>r.pos);
+    }
+    if r is NarrowRegister {
+        Register narrowed = r.underlying;
+        while narrowed is NarrowRegister {
+            narrowed = narrowed.underlying;
+        }
+        if !t:isSubtype(vc.typeContext(), r.semType, narrowed.semType) {
+            return vc.invalidErr("narrow register is not a subtype of narrowed register", <Position>r.pos);
+        }
```

Comment:
I don't think we need to check the every time we read form a NarrowRegister, just checking when we first narrow (ie: TypeMerge/TypeBr) is enough. That overlaps with my work.